src/Modules/OrchardCore.Commerce/Services/ProductListTitleFilterProvider.cs

```diff
@@ -0,0 +1,45 @@
+using OrchardCore.Commerce.Models;
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Records;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using YesSql;
+
+namespace OrchardCore.Commerce.Services;
+
+public class ProductListTitleFilterProvider : IProductListFilterProvider
+{
+    public const string TitleFilterId = "title";
+    public const string TitleAscOrderById = "TitleAsc";
+    public const string TitleDescOrderById = "TitleDesc";
+
+    public int Order { get; } = 10;
+
+    public Task<bool> CanHandleAsync(ProductListPart productList) => Task.FromResult(true);
+
+    public Task<IEnumerable<string>> GetOrderByOptionIdsAsync(ProductListPart productList) =>
+        Task.FromResult<IEnumerable<string>>(new[] { TitleAscOrderById, TitleDescOrderById });
+
+    public Task<IEnumerable<string>> GetFilterIdsAsync(ProductListPart productListPart) =>
+        Task.FromResult<IEnumerable<string>>(new[] { TitleFilterId });
+
+    public Task<IQuery<ContentItem>> BuildQueryAsync(ProductListFilterContext context)
+    {
+        var query = context.Query;
+        if (context.FilterParameters.FilterValues.TryGetValue(TitleFilterId, out var title))
+        {
+            query = query.With<ContentItemIndex>(index => index.DisplayText.Contains(title));
+        }
+
+        if (context.FilterParameters.OrderBy.Contains(TitleAscOrderById))
+        {
+            query = query.With<ContentItemIndex>().OrderBy(index => index.DisplayText);
+        }
+        else if (context.FilterParameters.OrderBy.Contains(TitleDescOrderById))
+        {
+            query = query.With<ContentItemIndex>().OrderByDescending(index => index.DisplayText);
+        }
```

Comment:
It's an interesting problem. First, I was aware that on the architectural level, it supports multiple orderbys, but on the UI, it doesn't. I wanted to keep it open for such a feature. Second, I intentionally have the providers set the orderby based on their Order parameter; is this a problem? This is what happens right now even if and `IList<string>` contains the active order by filters.
Regarding separate methods, I was thinking about it, too, however the `IQuery` API supports putting these (filters and orderbys) in any order so I didn't want to force doing it separately. Also, I can't even force it even with separate methods, because since you are working with the `IQuery` object, you have the ability to put orderbys to it from the Filter-specific method, too.

So what do you think, what should I change here?

Reply:
> Second, I intentionally have the providers set the orderby based on their Order parameter; is this a problem?

Yes, if we are to support multi-tiered sorting then the `˙IProductListFilterProvider.Order` shouldn't be a constant value like it is here. Because that means the precedence of the sorting criteria is set in stone. For example imagine if we also have a price filter provider with the Order = 20; but the user wants `?product.OrderBy=priceAsc&product.OrderBy=titleAsc` to get the cheapest products sorted by name when the price is the same. With the current approach it would still sort by title first and then by price.
One solution is to change `GetOrderedApplicableProvidersAsync` to sort by their index in `context.FilterParameters.OrderBy` instead of the Order property. (provider precedence between providers that only filter but not sort shouldn't matter) Although that would only work correctly as long as all filter providers only represent one order by option.

> Regarding separate methods, I was thinking about it, too, however the IQuery API supports putting these (filters and orderbys) in any order so I didn't want to force doing it separately. Also, I can't even force it even with separate methods, because since you are working with the IQuery object, you have the ability to put orderbys to it from the Filter-specific method, too.

It's more about encouraging the correct behavior than forcing it. If the dev really wants to mix those method calls hard enough, you can't really stop them even if you wanted to. But it would be nice to treat these two separately, and also I think mixing multiple wheres with orderbys in between is probably bad for performance and it wouldn't actually work with multiple sorting anyway.

Also another thing: you used `.OrderBy()` and `.OrderByDescending()`, but I think this will overwrite the sorting on subsequent calls? At least I think so, because we also have `.ThenBy()` and `.ThenByDescending()` which you are normally supposed to use on 2nd tier sorting. So I think this will not work either way. But if you separate the sorting into its own method and make it indicate if sorting happened, then you can select between OrderBy or ThenBy reliably.

Reply:
I think in this case the easiest solution is to not support multiple sorting options. I can't recall any webshops where it is possible. So I'll just set it a `string` property instead of `IList`. What do you think?

Reply:
Yeah, it's definitely not worth the effort.